apm-agent-plugins/apm-es-restclient-plugin/apm-es-restclient-plugin-common/src/main/java/co/elastic/apm/agent/es/restclient/ElasticsearchRestClientInstrumentationHelperImpl.java

```diff
@@ -0,0 +1,145 @@
+
+    public ElasticsearchRestClientInstrumentationHelperImpl(ElasticApmTracer tracer) {
+        this.tracer = tracer;
+        responseListenerObjectPool = QueueBasedObjectPool.ofRecyclable(
+            AtomicQueueFactory.<ResponseListenerWrapper>newQueue(createBoundedMpmc(MAX_POOLED_ELEMENTS)),
+            false,
+            new ResponseListenerAllocator());
+    }
+
+    private class ResponseListenerAllocator implements Allocator<ResponseListenerWrapper> {
+        @Override
+        public ResponseListenerWrapper createInstance() {
+            return new ResponseListenerWrapper(ElasticsearchRestClientInstrumentationHelperImpl.this);
+        }
+    }
+
+    @Override
+    @Nullable
+    public Span createClientSpan(String method, String endpoint, @Nullable HttpEntity httpEntity) {
+        final TraceContextHolder<?> activeSpan = tracer.getActive();
+        if (activeSpan == null || !activeSpan.isSampled()) {
+            return null;
+        }
+
+        // Don't record nested spans. In 5.x clients the instrumented sync method is calling the instrumented async method
+        if (activeSpan instanceof Span && ELASTICSEARCH.equals(((Span) activeSpan).getSubtype())) {
+            return null;
+        }
+
+        Span span = activeSpan.createSpan()
+            .withType(SPAN_TYPE)
+            .withSubtype(ELASTICSEARCH)
+            .withAction(SPAN_ACTION)
+            .appendToName("Elasticsearch: ").appendToName(method).appendToName(" ").appendToName(endpoint);
+        span.getContext().getDb().withType(ELASTICSEARCH);
+        span.activate();
+
+        if (span.isSampled()) {
+            span.getContext().getHttp().withMethod(method);
+            if (endpoint.endsWith(SEARCH_QUERY_PATH_SUFFIX)) {
+                if (httpEntity != null && httpEntity.isRepeatable()) {
+                    try {
+                        IOUtils.readUtf8Stream(httpEntity.getContent(), span.getContext().getDb().withStatementBuffer());
+                    } catch (IOException e) {
+                        // We can't log from here
+                    }
+                }
+            }
+        }
+        return span;
+    }
+
+    @Override
+    public void finishClientSpan(@Nullable Response response, Span span, @Nullable Throwable t) {
+        try {
+            String url = null;
+            int statusCode = -1;
+            if (response != null) {
+                url = response.getHost().toURI();
+                statusCode = response.getStatusLine().getStatusCode();
+            } else if (t != null) {
+                if (t instanceof ResponseException) {
+                    ResponseException esre = (ResponseException) t;
+                    url = esre.getResponse().getHost().toURI();
+                    statusCode = esre.getResponse().getStatusLine().getStatusCode();
+
+                        /*
+                        // Add tags so that they will be copied to error capture
```

Comment:
I would prefer to remove this and rather create a task to enhance the intake API.

Reply:
I will, maybe just leave a comment to remember that `esre.getResponse().getStatusLine().getReasonPhrase()` provides a useful info.